hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestListBlobProducer.java

```diff
@@ -59,65 +66,75 @@ public void setup() throws Exception {
   @Test
   public void testProducerWaitingForConsumerLagToGoDown() throws Exception {
     Configuration configuration = Mockito.spy(getRawConfiguration());
-    configuration.set(FS_AZURE_MAX_CONSUMER_LAG, "10");
+    configuration.set(FS_AZURE_PRODUCER_QUEUE_MAX_SIZE, "10");
     AzureBlobFileSystem fs = (AzureBlobFileSystem) FileSystem.newInstance(
         configuration);
     AbfsClient client = fs.getAbfsClient();
     AbfsClient spiedClient = Mockito.spy(client);
     fs.getAbfsStore().setClient(spiedClient);
     fs.setWorkingDirectory(new Path("/"));
     fs.mkdirs(new Path("/src"));
+    ExecutorService executor = Executors.newFixedThreadPool(5);
+    List<Future> futureList = new ArrayList<>();
     for (int i = 0; i < 20; i++) {
-      fs.create(new Path("/src/file" + i));
+      int iter = i;
+      futureList.add(executor.submit(() -> {
+        return fs.create(new Path("/src/file" + iter));
+      }));
+    }
+    for(Future future : futureList) {
+      future.get();
     }
-    AtomicBoolean produced = new AtomicBoolean(true);
 
     AtomicInteger producedBlobs = new AtomicInteger(0);
     AtomicInteger listBlobInvoked = new AtomicInteger(0);
 
+    final ITestListBlobProducer testObj = this;
+    final ListBlobQueue queue = new ListBlobQueue(
+        fs.getAbfsStore().getAbfsConfiguration().getProducerQueueMaxSize(),
+        1);
+
     Mockito.doAnswer(answer -> {
-          listBlobInvoked.incrementAndGet();
-          AbfsRestOperation op = client.getListBlobs(answer.getArgument(0),
-              answer.getArgument(1), 1, answer.getArgument(3));
-          producedBlobs.incrementAndGet();
-          produced.set(true);
-          return op;
+      synchronized (testObj) {
```

Comment:
Want to control that either of two blocks processed in totality and there be no context-switching from one block to the other:
```
 listBlobInvoked.incrementAndGet();
        AbfsRestOperation op = client.getListBlobs(answer.getArgument(0),
            answer.getArgument(1), 1, answer.getArgument(3));
        producedBlobs.incrementAndGet();
        if(producedBlobs.get() > 10) {
          Assert.assertTrue(queue.availableSize() > 0);
        }
        return op;
```
and
```
 consumer.consume();
  Assert.assertTrue(queue.availableSize() > 0);
```
This is to prevent in code-interleaving between the two block.
One example where inconsistency can happen:
1. `consumer.consume()` happens.
2. context-switch happened and the listing code got executed and also it populated the queue.
3. context-switch happens and now `queue.availableSize()` would give 0 which would fail the tests.

Need to ensure that when following runs, no other thread can change state of queue.
```
 consumer.consume();
  Assert.assertTrue(queue.availableSize() > 0);
```